src/BSML/FloatingScreen/FloatingScreenHandle.cpp

```diff
@@ -1,31 +1,143 @@
-        renderer->set_material(originalMaterial);
+        _isHovering = false;
+        UpdateMaterial();
+    }
+
+    void FloatingScreenHandle::OnPointerUp(UnityEngine::EventSystems::PointerEventData* eventData) {
+        auto currentEventSystem = EventSystems::EventSystem::get_current();
+        if (!currentEventSystem) return;
+        auto vrInputModule = currentEventSystem->get_currentInputModule().try_cast<VRUIControls::VRInputModule>().value_or(nullptr);
+        if (!vrInputModule) return;
+
+        _grabbingController = nullptr;
+        auto pointer = vrInputModule->_vrPointer;
+        _floatingScreen->OnHandleReleased(pointer);
+        
+        UpdateMaterial();
+    }
+
+    void FloatingScreenHandle::OnPointerDown(UnityEngine::EventSystems::PointerEventData* eventData) {
+        auto raycast = eventData->pointerPressRaycast;
+
+        auto go = raycast.get_gameObject();
+        if (go != this->get_gameObject()) return;
+
+        auto currentEventSystem = EventSystems::EventSystem::get_current();
+        if (!currentEventSystem) return;
+        auto vrInputModule = currentEventSystem->get_currentInputModule().try_cast<VRUIControls::VRInputModule>().value_or(nullptr);
+        if (!vrInputModule) return;
+
+        auto vrPointer = vrInputModule->_vrPointer;
+        auto vrController = vrPointer->_lastSelectedVrController;
+        if (!vrController) return;
+
+        _grabbingController = vrController;
+        _grabPos = vrController->_viewAnchorTransform->InverseTransformPoint(
+                _floatingScreen->get_transform()->get_position()
+        );
+        _grabRot = Quaternion::Inverse(vrController->_viewAnchorTransform->get_rotation()) * _floatingScreen->get_transform()->get_rotation();
+
+        _floatingScreen->OnHandleGrab(vrPointer);
+
+        UpdateMaterial();
+    }
+
+    void FloatingScreenHandle::Init(FloatingScreen* floatingScreen) {
+        _floatingScreen = floatingScreen;
+    }
+
+    void FloatingScreenHandle::Update() {
+        // If we're not grabbing, don't do anything
+        if (!_grabbingController) return;
+
+        _grabPos -= Vector3::get_forward() * (-_grabbingController->get_thumbstick().y * Time::get_unscaledDeltaTime());
+        Vector3 targetPosition = _grabbingController->_viewAnchorTransform->get_transform()->TransformPoint(_grabPos);
+        Quaternion targetRotation = _grabbingController->_viewAnchorTransform->get_transform()->get_rotation() * _grabRot;
+
+        _floatingScreen->get_transform()->SetPositionAndRotation(
+            Vector3::Lerp(_floatingScreen->get_transform()->get_position(), targetPosition, 10.0f * Time::get_unscaledDeltaTime()),
+            Quaternion::Lerp(_floatingScreen->get_transform()->get_rotation(), targetRotation, 5.0f * Time::get_unscaledDeltaTime())
+        );
+    }
+    
+    void FloatingScreenHandle::OnDestroy() {
+        _grabbingController = nullptr;
+        _floatingScreen = nullptr;
+    }
```

Comment:
I'm not sure if we even need to clear those on destroy

Reply:
probably not needed but it doesn't matter all that much